core/trie2/triedb/pathdb/layertree_test.go

```diff
@@ -0,0 +1,495 @@
+}
+
+// mockTrieNode is a simple implementation of trienode.TrieNode for testing
+type mockTrieNode struct {
+	blob []byte
+}
+
+func (m *mockTrieNode) Blob() []byte    { return m.blob }
+func (m *mockTrieNode) Hash() felt.Felt { return *new(felt.Felt).SetBytes(m.blob) }
+func (m *mockTrieNode) IsLeaf() bool    { return len(m.blob) == 0 }
+
+// layerTracker tracks trie nodes across different layers to simplify testing
+type layerTracker struct {
+	// Mimic the layer tree by mapping the state root to the nodes
+	classNodes           map[felt.Felt]map[trieutils.Path]trienode.TrieNode
+	contractNodes        map[felt.Felt]map[trieutils.Path]trienode.TrieNode
+	contractStorageNodes map[felt.Felt]map[felt.Felt]map[trieutils.Path]trienode.TrieNode
+
+	// Tracks all unique and latest nodes in the layer tree
+	classPaths           map[trieutils.Path]trienode.TrieNode
+	contractPaths        map[trieutils.Path]trienode.TrieNode
+	contractStoragePaths map[felt.Felt]map[trieutils.Path]trienode.TrieNode
+
+	// Child to parent layer relationship
+	childToParent map[felt.Felt]felt.Felt
+}
+
+func newLayerTracker() *layerTracker {
+	return &layerTracker{
+		classNodes:           make(map[felt.Felt]map[trieutils.Path]trienode.TrieNode),
+		contractNodes:        make(map[felt.Felt]map[trieutils.Path]trienode.TrieNode),
+		contractStorageNodes: make(map[felt.Felt]map[felt.Felt]map[trieutils.Path]trienode.TrieNode),
+		childToParent:        make(map[felt.Felt]felt.Felt),
+		classPaths:           make(map[trieutils.Path]trienode.TrieNode),
+		contractPaths:        make(map[trieutils.Path]trienode.TrieNode),
+		contractStoragePaths: make(map[felt.Felt]map[trieutils.Path]trienode.TrieNode),
+	}
+}
+
+func (t *layerTracker) trackLayer(root, parent felt.Felt) {
+	if root.Equal(&parent) {
+		return
+	}
+	t.childToParent[root] = parent
+}
+
+func (t *layerTracker) trackNodes(
+	root felt.Felt,
+	parent felt.Felt,
+	classNodes,
+	contractNodes map[trieutils.Path]trienode.TrieNode,
+	storageNodes map[felt.Felt]map[trieutils.Path]trienode.TrieNode,
+) {
+	t.trackClassNodes(root, classNodes)
+	t.trackContractNodes(root, contractNodes)
+	t.trackContractStorageNodes(root, storageNodes)
+	t.trackLayer(root, parent)
+}
+
+func (t *layerTracker) trackClassNodes(root felt.Felt, nodes map[trieutils.Path]trienode.TrieNode) {
+	for path, node := range nodes {
+		if t.classNodes[root] == nil {
+			t.classNodes[root] = make(map[trieutils.Path]trienode.TrieNode)
+		}
+		t.classNodes[root][path] = node
+		t.classPaths[path] = node
+	}
+}
+
+func (t *layerTracker) trackContractNodes(root felt.Felt, nodes map[trieutils.Path]trienode.TrieNode) {
+	for path, node := range nodes {
+		if t.contractNodes[root] == nil {
+			t.contractNodes[root] = make(map[trieutils.Path]trienode.TrieNode)
+		}
+		t.contractNodes[root][path] = node
+		t.contractPaths[path] = node
+	}
+}
+
+func (t *layerTracker) trackContractStorageNodes(root felt.Felt, nodes map[felt.Felt]map[trieutils.Path]trienode.TrieNode) {
```

Comment:
I'll just mark here onwards that this function should receive values by reference. In this case is `root`

Reply:
done